File: misc/missing_number/missing_number.py
def find_missing(numbers):
    """
    >>> find_missing([1,2,3,4,5,6,7,8])
    9
    >>> find_missing([2,3,4,5,6,7,8,9])
    1
    >>> find_missing([1,2,3,4,5,6,7,9])
    8
    >>> find_missing([1,3,4,5,6,7,8,9])
    2
    >>> find_missing([1,2,4,5,6,7,8,9])
    3
    >>> find_missing([1,3,4,5,6])
    2
    >>> find_missing([1,2,3,4,6])
    5
    >>> find_missing([1,2,4,5,6])
    3
    """
    left = 0
    right = len(numbers) - 1

    while left < right:
        middle = (left + right) // 2
        if numbers[middle] != middle + 1:
            if right == middle:
                return middle + 1
            right = middle
        else:
            if left == middle:
                if numbers[right] == right + 1:
                    return right + 2
                return right + 1
            left = middle
    return 1

File: misc/missing_number/missing_number.exs
defmodule FindMissing do
  def for(items) do
    last = tuple_size(items) - 1
    _middle(items, div(last, 2), 0, last)
  end

  defp _middle(items, middle, first, last) 
    when elem(items, middle) === middle + 1 and middle > first,
    do: _middle(items, div(last + middle, 2), middle, last)
  defp _middle(items, middle, first, last) 
    when elem(items, middle) !== middle + 1 and middle < last,
    do: _middle(items, div(first + middle, 2), first, middle)

  defp _middle(items, _, _, last)
    when elem(items, last) === last + 1,
    do: last + 2
  defp _middle(_, _, _, last), do: last + 1
end

IO.puts FindMissing.for({2,3,4,5,6,7,8,9})
IO.puts FindMissing.for({1,3,4,5,6,7,8,9})
IO.puts FindMissing.for({1,2,4,5,6,7,8,9})
IO.puts FindMissing.for({1,2,3,5,6,7,8,9})
IO.puts FindMissing.for({1,2,3,4,6,7,8,9})
IO.puts FindMissing.for({1,2,3,4,5,7,8,9})
IO.puts FindMissing.for({1,2,3,4,5,6,8,9})
IO.puts FindMissing.for({1,2,3,4,5,6,7,9})
IO.puts FindMissing.for({1,2,3,4,5,6,7,8})
